### PR TITLE
switch hash function to murmur3

### PR DESCRIPTION
**Pull request by ianopolous**
(And add test to guard against trivial hash prefix collisions)

### CODE REVIEW

File: hamt.go [hunk 1]
```diff
@@ -3,10 +3,10 @@ package hamt
 import (
 	"context"
 	"fmt"
-	"hash/fnv"
 	"math/big"
 
 	cid "github.com/ipfs/go-cid"
+	"gx/ipfs/QmfJHywXQu98UeZtGJBQrPAR6AtmDjjbe3qjTo9piXHPnx/murmur3"
```
Comment (Kubuxu) [inline hamt.go]: You need to add it to package.json and keep it un-rewritten.

Comment (ianopolous) [reply hamt.go]: I just copied it from go-ipfs. What is the original?

Comment (Kubuxu) [reply hamt.go]: Entry: https://github.com/ipfs/go-ipfs/blob/master/package.json#L303-L308

https://ipfs.io/ipfs/QmfJHywXQu98UeZtGJBQrPAR6AtmDjjbe3qjTo9piXHPnx/murmur3/package.json

**Comment by Kubuxu** [comment]
I also think it might be better to keep 128 bit wide hash function there. The maximum depth of the tire depends on size of the hash function.

**Comment by whyrusleeping** [comment]
> The maximum depth of the tire depends on size of the hash function.

Yeah, I was thinking of doing something to extend this further (once you run out, rehash with the key concatenated by itself N times) but that seemed like a lot of extra work.

**Comment by whyrusleeping** [comment]
Also note that finding a hash collision doesnt actually increase the depth until you have four collisions at a given depth.

**Comment by ianopolous** [comment]
@whyrusleeping yep, but with the existing hash function the collision is then n layers deep, where n is the largest common prefix of those 4 keys

**Comment by ianopolous** [comment]
@whyrusleeping another possibility to remove the depth limit would be to hash again after passing multiple of the hash size.

**Comment by whyrusleeping** [comment]
@ianopolous just hashing again wouldnt work, anything thats going to become an issue that far down is going to be a hash collision on the 128 bit keyspace. hashing the same keys will continue with the collision. Hashing the key concatenated with itself would (hopefully) break the collision.

**Comment by ianopolous** [comment]
@whyrusleeping right you are! :-) another option would be to switch to a cryptographic hash function after running out of hash length. Then a collision would be news.

**Pull request merged**